File: csharp/spiral-matrix/SpiralMatrix.cs
public class SpiralMatrix
{
    public static int[,] GetMatrix(int size)
    {
        var matrix = new int[size, size];

        var top = 0;
        var bottom = size - 1;
        var left = 0;
        var right = size - 1;

        var num = 1;

        while (top <= bottom && left <= right)
        {
            for (var col = left; col <= right; col++)
            {
                matrix[top, col] = num++;
            }
            top++;

            for (var row = top; row <= bottom; row++)
            {
                matrix[row, right] = num++;
            }
            right--;

            if (top <= bottom)
            {
                for (var col = right; col >= left; col--)
                {
                    matrix[bottom, col] = num++;
                }
                bottom--;
            }

            if (left <= right)
            {
                for (var row = bottom; row >= top; row--)
                {
                    matrix[row, left] = num++;
                }
                left++;
            }
        }

        return matrix;
    }
}

File: cpp/spiral-matrix/spiral_matrix.cpp
#include "spiral_matrix.h"

namespace spiral_matrix
{

    std::vector<std::vector<uint32_t>> spiral_matrix(uint32_t size)
    {
        std::vector<std::vector<uint32_t>> matrix(
            size,
            std::vector<uint32_t>(size, 0));

        int top = 0;
        int bottom = static_cast<int>(size) - 1;
        int left = 0;
        int right = static_cast<int>(size) - 1;

        uint32_t num = 1;

        while (top <= bottom && left <= right)
        {
            for (int col = left; col <= right; col++)
            {
                matrix[top][col] = num++;
            }
            top++;

            for (int row = top; row <= bottom; row++)
            {
                matrix[row][right] = num++;
            }
            right--;

            if (top <= bottom)
            {
                for (int col = right; col >= left; col--)
                {
                    matrix[bottom][col] = num++;
                }
                bottom--;
            }

            if (left <= right)
            {
                for (int row = bottom; row >= top; row--)
                {
                    matrix[row][left] = num++;
                }
                left++;
            }
        }

        return matrix;
    }

} // namespace spiral_matrix
